core/server/master/src/main/java/alluxio/master/file/MetadataSyncLock.java

```diff
@@ -0,0 +1,68 @@
+package alluxio.master.file;
+
+import alluxio.AlluxioURI;
+
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class MetadataSyncLock {
+  private static Map<String, ReentrantReadWriteLock> sLockMap = new ConcurrentHashMap<>();
+
+  /*
+    e.g. for path /a/b/c/d
+    1. read lock /
+    2. read lock /a/
+    3. read lock /a/b/c
+    4. write lock /a/b/c/d
+   */
+  public static void lockPath(AlluxioURI uri) {
+    int depth = uri.getDepth();
+    // TODO(yiminwei) potential inefficient string operations
+    for (int i = 0; i <= depth; ++i) {
+      String lockKey = uri.getLeadingPath(i);
+      ReentrantReadWriteLock lock = sLockMap.computeIfAbsent(lockKey, (k) -> new ReentrantReadWriteLock());
+      if (i != depth) {
+        System.out.println("Read Lock" + lockKey);
+        lock.readLock().lock();
+      } else {
+        System.out.println("Write Lock" + lockKey);
+        lock.writeLock().lock();
+      }
+    }
+  }
+
+  /*
+    e.g. for path /a/b/c/d
+    1. write unlock /a/b/c/d
+    2. read unlock /a/b/c
+    3. read unlock /a/
+    4. read unlock /
+   */
+  public static void unlockPath(AlluxioURI uri) {
+    int depth = uri.getDepth();
+    // TODO(yiminwei) potential inefficient string operations
+    for (int i = depth; i >= 0; --i) {
+      String lockKey = uri.getLeadingPath(i);
+      int finalI = i;
+      sLockMap.computeIfPresent(lockKey, (k, v) -> {
+        if (finalI == depth) {
+          System.out.println("Write Unlock" + lockKey);
+          v.writeLock().unlock();
+        } else {
+          System.out.println("Read Unlock" + lockKey);
+          v.readLock().unlock();
+        }
+        if (!v.isWriteLocked() && v.getReadLockCount() == 0) {
+          // garbage collect lock not being used to prevent memory overflow
+          // this can be converted to a timer based clean up
+          System.out.println("Destroy lock " + k);
+          return null;
```

Comment:
There is a data race here with the lockPath function. The lockPath function may not actually performing the locking operations until after the lock is removed from the map.

Reply:
Actually there's already something dedicated for read write lock assignment and garbage collection. It's called [LockPool](https://github.com/Alluxio/alluxio/blob/master/core/common/src/main/java/alluxio/collections/LockPool.java) in alluxio so I just turned to use it.